packages/api/src/routes/devices/internal-user.ts

```diff
@@ -0,0 +1,66 @@
+import { Request, Response } from "express";
+import Router from "express-promise-router";
+import status from "http-status";
+import { asyncHandler } from "../util";
+import {
+  getAllConnectedUsers,
+  getConnectedUserOrFail,
+} from "../../command/connected-user/get-connected-user";
+import { Constants, providerOAuth2OptionsSchema } from "../../shared/constants";
+import { updateProviderData } from "../../command/connected-user/save-connected-user";
+import { sendProviderDisconnected } from "../../command/webhook/devices";
+import { getUUIDFrom } from "../schemas/uuid";
+
+const router = Router();
+
+/** ---------------------------------------------------------------------------
+ * POST /internal/user/refresh-tokens
+ *
+ * Attempts to refresh all users' tokens for each of their connected providers.
+ * If the refresh doesn't work, this means that the token is invalid and can no
+ * longer be used - in this case, the token will be revoked locally and a
+ * webhook will be sent to the CX to notify them of the disconnect.
+ *
+ * @param req.query.cxId - (Optional) the customer/account's ID.
+ * @return 200 OK
+ */
+router.post(
+  "/refresh-tokens",
+  asyncHandler(async (req: Request, res: Response) => {
+    const cxId = getUUIDFrom("query", req, "cxId").optional();
+    const connectedUsers = await getAllConnectedUsers(cxId);
+    let disconnectCount = 0;
+    for (const connectedUser of connectedUsers) {
+      const providers = connectedUser.providerMap ? Object.keys(connectedUser.providerMap) : [];
+      const disconnectedProviders: string[] = [];
+      for (const providerStr of providers) {
+        const providerOAuth2Type = providerOAuth2OptionsSchema.safeParse(providerStr);
+        if (providerOAuth2Type.success) {
+          const providerOAuth2 = Constants.PROVIDER_OAUTH2_MAP[providerOAuth2Type.data];
+          try {
+            await providerOAuth2.getAccessToken(connectedUser);
+          } catch (error) {
+            await updateProviderData({
+              id: connectedUser.id,
+              cxId: connectedUser.cxId,
+              provider: providerOAuth2Type.data,
+              providerItem: undefined,
+            });
+            disconnectedProviders.push(providerStr);
+          }
+        }
+      }
+      if (disconnectedProviders) {
+        const newConnectedUser = await getConnectedUserOrFail({
+          cxId: connectedUser.cxId,
+          id: connectedUser.id,
+        });
+        await sendProviderDisconnected(newConnectedUser, disconnectedProviders);
+        disconnectCount++;
+      }
+    }
+    return res.send(status.OK).json({ usersProcessed: connectedUsers.length, disconnectCount });
```

Comment:
nit: `disconnectCount` is ambiguous for the person calling the endpoint; "does it mean the amount of disconnected providers or the amount of users with at least one disconnected provider?"

Maybe could rename to indicate is the latter?